src/runtime/tests/manifest-parser-test.ts

```diff
@@ -538,6 +538,76 @@ describe('manifest parser', () => {
         )
     `);
   });
+  it('parses type variables without constraints', () => {
+    parse(`
+      particle Foo
+        data: reads ~a
+    `);
+  });
+  it('parses type variables with a constraint', () => {
+    parse(`
+      particle Foo
+        data: reads ~a with {name: Text}
+    `);
+  });
+  it('parses type variables with multiple constraints', () => {
+    parse(`
+      particle Foo
+        data: reads ~a with {name: Text, age: Number}
+    `);
+  });
+  it('parses max type variables without constraints', () => {
+    parse(`
+      particle Foo
+        data: reads ~a with {*}
+    `);
+  });
+  it('parses max type variables with a constraint', () => {
+    parse(`
+      particle Foo
+        data: reads ~a with {name: Text, *}
+    `);
+  });
+  it('parses max type variables with multiple constraints', () => {
+    parse(`
+      particle Foo
+        data: reads ~a with {name: Text, age: Number, *}
+    `);
+  });
+  it.only('fails to parse max type variables with oddly-ordered constraints', () => {
+    assert.throws(() => {
+      parse(`
+        particle Foo
+          data: reads ~a with {name: Text, *, age: Number}
+      `);
+    });
+    assert.throws(() => {
+      parse(`
+        particle Foo
+          data: reads ~a with {*, age: Number}
+      `);
+    });
+  });
+  it('fails to parse max type variables with multiple `*`s', () => {
```

Comment:
I think this should be a warning, not a parse failure (otherwise it is too restrictive and could slow developers down).